### DISCUSSION TITLE
`Discussion`: `Compiler` - behavioural changes testing/dogfooding

### DISCUSSION THREAD

**Discussion by vzarytovskii** [dotnet/fsharp#16521] [Ideas]
After chatting a bit with @dsyme, team, and observing community suggestions, I propose we start introducing behavioural/codegen changes for **testing by external community/customers** via compiler flags (and requiring `langversion:preview`), for example:

* We can introduce a new codegen/optimizer flag (e.g. `--optimize:devirtualize_equality`), which will be devirtualizing certain calls (in case of equality, or comparison), resulting in better non-boxing codegen for value types.

The reason for having those - we want to be able to let users test certain changes, and provide feedback, but not tie them to language version (or make it default for the time being) without risk of breaking existing customers.

**Comment by smoothdeveloper** [comment]
I don't know how pressing the need is, but if there is room for infrastructure around it, should this be handled by language pragma (and having support for those to be unset)?

https://github.com/fsharp/fslang-suggestions/issues/278

**Reply by vzarytovskii** [reply]
Not sure what language pragmas mean, but if it's anything like Haskell features, we absolutely don't want them at this point.
Features will still be tied to language version. Some codegen or optimizer changes - their own flags.
